# artMS

___Analytical R Tools for Mass Spectrometry___

---

[![Travis build status](https://travis-ci.org/biodavidjm/artMS.svg?branch=master)](https://travis-ci.org/biodavidjm/artMS)
[![codecov](https://codecov.io/gh/biodavidjm/artMS/branch/master/graph/badge.svg)](https://codecov.io/gh/biodavidjm/artMS)


## Overview

`artMS` is an R package that provides a set of tools for the analysis and 
integration of large-scale proteomics datasets (mass-spectrometry-based) 
obtained using the popular proteomics software package 
[MaxQuant](http://www.biochem.mpg.de/5111795/maxquant).

`artMS` facilitates and simplifies the analysis on R by creating a 
configuration file where multiple options can be selected, including:

- Multiple quality control (QC) functions.
- Relative quantification using [MSstats](http://msstats.org/).
- Post-MSstats processing and integration.
- A set of miscellaneous functions to generate input files for other popular 
tools, including:
    - APMS: [SAINTq and SAINT express](http://saint-apms.sourceforge.net/Main.html)
    - [Photon](https://github.com/jdrudolph/photon)
    - [Phosfate](http://phosfate.com/)

***Required input files***

`artMS` performs a full set of analyses by taking the following files:

- `evidence.txt` file: The output of the quantitative proteomics software 
package `MaxQuant`. 
- `keys.txt` (tab-delimited) txt file describing the experimental designed 
(check below to learn how to create it).
- `contrast.txt` (tab-delimited) file with the comparisons between conditions 
that the user wished to be quantified (check below to learn how to create it).
- `config.yaml`: `yaml` file which enables the customization of the parameters for the quantification (and other operations, including QC analysis and extra plots and annotations). A configuration file template can be generated by running `artms_writeConfigYamlFile()`


## How to install

`artMS` is currently under revision by BioConductor. Until officially accepted, 
the development version can be installed directly from Github.

But before you start, make sure that you are running an `R version >= 3.5` or otherwise nothing will work (Bioconductor requirement). 
You can check the R version currently running on your system by running 
this command in RStudio:

```
getRversion()
```

If the outcome is `>= 3.5.0`, congratulations! you can move forward

*If it is not, then you need to [install the latest version of R in your system](https://www.r-project.org/).*

After updating to the latest R version, open RStudio and try again 
`getRversion()` to make sure it worked.

Once you have the latest `R (>= 3.5)` version running in your system, 
follow these steps:

- Install the following annotation packages. The installation of 
these packages returns multiple warnings that do not affect the functionality of `artMS` 
but could generate problems during the installation of `artMS`.
```
source("https://bioconductor.org/biocLite.R")
biocLite(c('org.Ag.eg.db', 'org.At.tair.db', 'org.Bt.eg.db', 'org.Ce.eg.db', 'org.Cf.eg.db', 'org.Dm.eg.db', 'org.Dr.eg.db', 'org.EcK12.eg.db', 'org.EcSakai.eg.db', 'org.Gg.eg.db', 'org.Hs.eg.db', 'org.Mm.eg.db', 'org.Mmu.eg.db', 'org.Pf.plasmo.db', 'org.Pt.eg.db', 'org.Rn.eg.db', 'org.Sc.sgd.db', 'org.Ss.eg.db', 'org.Xl.eg.db'))
```

- Install `artMS`:

```
install.packages("devtools")
library(devtools)
install_github("biodavidjm/artMS", build_vignettes=TRUE)
```

- Check that it is up and running by checking, for example, the documentation of the qc function `artms_qualityControlEvidenceBasic`:

```
library(artMS)
?artms_qualityControlEvidenceBasic
```

Once installed, we suggest you to do a quick test by running the quality 
control functions on the "evidence" (`artms_data_ph_evidence`) and 
"keys" (`artms_data_ph_keys`) files included in `artMS` as test datasets.

```
# First go to a working directory: several pdfs will be generated
setwd("/path/to/your/working/directory/")

# And run:
artms_qualityControlEvidenceBasic(evidence_file = artms_data_ph_evidence, keys_file = artms_data_ph_keys, prot_exp =  "PH")

artms_qualityControlEvidenceExtended(evidence_file = artms_data_ph_evidence, keys_file = artms_data_ph_keys)
```

Go to the folder `"/path/to/your/working/directory/"` and check out all the 
generated QC (pdf) files.

(To learn more about these testing datasets, check the documentation by typing 
`?artms_data_ph_keys` or `?artms_data_ph_evidence`)

## The `artMS` input files

Three basic tab-delimited files are required to perform the full pack of 
operations:

### `evidence.txt`

The output of the quantitative proteomics software package `MaxQuant`. 
It integrates all the information about the identified peptides.

### `keys.txt`

It contains the experimental design. 
The Keys will be merged with the `evidence.txt`
file through the "RawFile" column. 
Each RawFile corresponds to an unique individual technical replicate / 
biological replicate / Condition / Run.

Example: (`artMS` data object: `artms_data_ph_keys`)

**RawFile**|**IsotopeLabelType**|**Condition**|**BioReplicate**|**Run**
:-----:|:-----:|:-----:|:-----:|:-----:
FU20170922-17|L|H1N1\_03H|H1N1\_03H-1|9
FU20170922-19|L|H1N1\_03H|H1N1\_03H-2|10
FU20170922-21|L|H1N1\_06H|H1N1\_06H-1|11
FU20170922-23|L|H1N1\_06H|H1N1\_06H-2|12
FU20170922-35|L|H1N1\_12H|H1N1\_12H-1|13
FU20170922-37|L|H1N1\_12H|H1N1\_12H-2|14
FU20170922-39|L|H1N1\_18H|H1N1\_18H-1|15
FU20170922-41|L|H1N1\_18H|H1N1\_18H-2|16
FU20170922-01|L|MOCK\_03H|MOCK\_03H-1|1
FU20170922-03|L|MOCK\_03H|MOCK\_03H-2|2
FU20170922-05|L|MOCK\_06H|MOCK\_06H-1|3
FU20170922-07|L|MOCK\_06H|MOCK\_06H-2|4
FU20170922-09|L|MOCK\_12H|MOCK\_12H-1|5
FU20170922-11|L|MOCK\_12H|MOCK\_12H-2|6
FU20170922-13|L|MOCK\_18H|MOCK\_18H-1|7
FU20170922-15|L|MOCK\_18H|MOCK\_18H-2|8

**Requirement**: 
- Use only numbers (0 - 9) and the letters (A - Z, both uppercase and lowercase) for the Conditions' names. The only special character allowed for the condition names is underscore (`_`).
- A condition name CANNOT start by a number
- Use the same labeling for `BioReplicate` as the one used for `Condition`, but with an 
additional `dash (-)` to indicate the number of biological replicate. For example, if condition `H1N1_06H` has too biological replicates, name them as follows: `H1N1_06H-1` and  `H1N1_06H-2`

### `contrast.txt`

The comparisons between conditions that the user wants to quantify. The names of the comparisons must follow the following consensus:

```
Condition_A-Condition_B_mutant
```

- The two conditions to be compared must be separated by a dash symbol (`-`)

As a result of the quantification, the condition on the left will take the positive log2FC 
sign (if the protein is more abundant in condition `Condition_A`) and the condition 
on the right the negative log2FC (if a protein is more abundant in condition 
`Condition_B_mutant`)

Example:

```
H1N1_03H-MOCK_03H
H1N1_06H-MOCK_06H
H1N1_12H-MOCK_12H
H1N1_18H-MOCK_18H
```


## The configuration file (`.yaml`)

The configuration file (`yaml` format) contains the configuration details for
the quantification performed by `artMS` using `MSstats`. 

To generate a sample configuration file, go to your project folder and run:

```
artms_writeConfigYamlFile(config_file_name = "config.yaml" )
```

`artMS` can quantify changes protein abundance, 
phosphorylation (ph), ubiquitination (ub), and acetylation (ac).

*Although it seems complex, the default options work very well*. 

The configuration (`yaml`) file contains the following sections:

### `files`

```
files :
  evidence : /path/to/the/evidence.txt
  keys : /path/to/the/keys.txt
  contrasts : /path/to/the/contrast.txt
  output : /path/to/the/results_folder/results.txt
```

The file `path/name` of the required files. It is recommended to create a new folder in your folder project (for example, `results_folder`). The results file name (e.g. `results.txt`) will be used as prefix for the several files (`txt` and `pdf`) that will be generated.

---

### `qc`

```
qc:
  basic: 1 # 1 = yes; 0 = no
  extended: 1 # 1 = yes; 0 = no
```

- basic: if chosen (1), it will generate the following pdf files:
      - `-basicReproducibility.pdf`: correlation dot plot for all the combinations of biological replicates of conditions, based on MS Intensity values using features (peptide+charge)
      - `-correlationMatrixBR.pdf`: It contains 3 pages. *Correlation matrix* for all the biological replicates using MS Intensity values, *Clustering matrix* of the MS Intensities and correlation distribution *histogram*.
      - `-correlationMatrixBR.pdf`: Same as the previous one, but based on MS Intensity values of Conditions
      - `-IntensityDistributions.pdf`: 2 pages. *Box-dot plot* and *Jitter plot* of biological replicates based on MS (raw) intensity values.
      - `-intensityStats.pdf`: several pages, including bar plots of *Total Sum of Intensities in BioReplicates*, *Total Sum of Intensities in Conditions*, *Total Peptide Counts in BioReplicates*, *Total Peptide Counts in conditions* separated by categories (`CON`: contaminants, `PROT` peptides, `REV` reversed sequences); *Box plots* of MS Intensity values per biological replicates and conditions; *bar plots* of total intensity (excluding contaminants) by bioreplicates and conditions; Barplots of *total feature counts* by bioreplicates and conditions.
      - `-ptmStats.pdf`: If any PTM is selected (`PH`, `UB`, `AC`) an extra pdf file will be generated with stats related to the selected modification, including: *bar plot of peptide counts and intensities*, broken by `PTM/other` categories; bar plots of *total sum-up of MS intensity values* by other/PTM categories.
- extended: if chosen (1), it will generate the following QC files:
      - `QC-ID-Overlap.pdf`
      - `QC-IntCorrelation.pdf`
      - `QC-SamplePrep.pdf`
      - `QC_Plots_CHARGESTATE.pdf`
      - `QC_Plots_IONS.pdf`
      - `QC_Plots_MASSERROR.pdf`
      - `QC_Plots_MZ.pdf`
      - `QC_Plots_PepDetect.pdf`
      - `QC_Plots_PEPINT.pdf`
      - `QC_Plots_PepIonOversampling.pdf`
      - `QC_Plots_PEPTIDES.pdf`
      - `QC_Plots_ProtDetect.pdf`
      - `QC_Plots_PROTEINS.pdf`
      - `QC_Plots_ProtInt.pdf`
      - `QC_Plots_PSM.pdf`
      - `QC_Plots_summary_ISOTOPE.pdf`
      - `QC_Plots_summary_MS1SCANS.pdf`
      - `QC_Plots_summary_MS2.pdf`
      - `QC_Plots_summary_MSMS.pdf`
      - `QC_Plots_TYPE.pdf`

---

### `data`

```
data:
  enabled : 1 # 1 = yes; 0 = no
  fractions: 
    enabled : 0 # 1 for protein fractionation
  silac: 
    enabled : 0 # 1 for SILAC experiments
  filters: 
    enabled : 1
    contaminants : 1
    protein_groups : remove #remove, keep
    modifications : AB # PH, UB, AB, APMS
  sample_plots : 1 # correlation plots
```

Let's break it down `data`:

- `enabled`:
      - `1`: to pre-process the data provided in the *files* section.
      - `0`: won't process the data (and a pre-generated MSstats file will be expected)

- `fractions`: Multiple fractionation or separation methods are often combined 
in proteomics to improve signal-to-noise and proteome coverage and to 
reduce interference between peptides in quantitative proteomics.
      - `enabled : 1` for fractionation dataset
      - `enabled : 0` no fractions

- `silac`:
    - `enabled : 1`: if the files belong to a SILAC experiment. See **Special case: SILAC** below for details
    - `enabled : 0`: it does not

- `filters`: the following filtering options are available
    -  `enabled : 1` Enables filtering
    -  `contaminants : 1` Removes contaminants (`CON__` and `REV__` labeled by MaxQuant)
    - `protein_groups : ` choose whether `remove` or `keep` protein groups
    -  `modifications : ` any of the proteomics experiments, `PH`, `UB`, or `AC` for posttranslational modifications, `AB` or `APMS` otherwise.

- `sample_plots` 
    - `1` Generate correlation plots
    - `0` otherwise

---

### `msstats`

```
msstats :
  enabled : 1
  msstats_input : 
  profilePlots : none # before, after, before-after (recommended), none
  normalization_method : equalizeMedians # globalStandards (include a reference protein(s) ), equalizeMedians, quantile, 0
  normalization_reference :  # if globalStandards is chosen, then an UNIPROT ID must be provided.
  summaryMethod : TMP # "TMP"(default) means Tukey's median polish, which is robust estimation method. "linear" uses linear mixed model. "logOfSum" conducts log2 (sum of intensities) per run.
  censoredInt : NA  # Missing values are censored or at random. 'NA' (default) assumes that all 'NA's in 'Intensity' column are censored. '0' uses zero intensities as censored intensity. In this case, NA intensities are missing at random. The output from Skyline should use '0'. Null assumes that all NA intensites are randomly missing.
  cutoffCensored : minFeature  # Cutoff value for censoring. only with censoredInt='NA' or '0'. Default is 'minFeature', which uses minimum value for each feature.'minFeatureNRun' uses the smallest between minimum value of corresponding feature and minimum value of corresponding run. 'minRun' uses minumum value for each run.
  MBimpute : 1 # only for summaryMethod="TMP" and censoredInt='NA' or '0'. TRUE (default) imputes 'NA' or '0' (depending on censoredInt option) by Accelated failure model. FALSE uses the values assigned by cutoffCensored.
  feature_subset: all # all|highQuality  : highQuality seems to be buggy right now
```

Let's break it down:

- `enabled : ` Choose `1` to run MSstats, `0` otherwise.
- `msstats_input :` blank if MSstats is going to be run (`enabled : 1`). But if otherwise (`enabled : 0) then provide the path to the previously generated `evidence-mss.txt`
- `profilePlots :` Several profile plots available. 
    * `before` plots only before normalization
    * `after` plots only after normalization
    * `before-after`: recommended, although computational expensive (time consuming)
    * `none` no normalization plots (convenient if time limitations)

- `normalization_method :` available options:
    - `equalizeMedians`
    - `quantile`
    - `0`: no normalization (not recommended)
    - `globalStandards` if selected, specified the reference protein in `normalization_reference` (next)
- `normalization_reference :` an UniProt id if `globalStandards` is chosen as the `normalization_method` (above)
- `summaryMethod :` TMP # "TMP"(default) means Tukey's median polish, which is robust estimation method. "linear" uses linear mixed model. "logOfSum" conducts log2 (sum of intensities) per run.
- `censoredInt :` 
    - `NA`  Missing values are censored or at random. 'NA' (default) assumes that all 'NA's in 'Intensity' column are censored. 
    - `0` uses zero intensities as censored intensity. In this case, NA intensities are missing at random. The output from Skyline should use `0`. Null assumes that all `NA` intensities are randomly missing.
- `cutoffCensored :` 
    - `minFeature` Cutoff value for censoring. Only with `censoredInt='NA'` or `0`. Default is 'minFeature', which uses minimum value for each feature.
    - `minFeatureNRun` uses the smallest between minimum value of corresponding feature and minimum value of corresponding run. 
    - `minRun` uses minimum value for each run.
- `MBimpute :` 
    - `TRUE` only for `summaryMethod="TMP"` and `censoredInt='NA'` or `0`. TRUE (default) imputes 'NA' or '0' (depending on censoredInt option) by Accelerated failure model. 
    - `FALSE` uses the values assigned by cutoffCensored.
- `feature_subset :` 
    - `all` : default
    - `highQuality`  : this option seems to be buggy right now

Check [MSstats documentation](http://msstats.org/) to find out more about every option.

---

### `output_extras`

```
  enabled : 1 # if 0, won't do anything specified in this section
  annotate :  
    enabled: 1 # if 1, will generate a `-results-annotated.txt` file that including Gene and Protein.Name
    specie : HUMAN  # Supported species: HUMAN, MOUSE, ANOPHELES, ARABIDOPSIS, BOVINE, WORM, CANINE, FLY, ZEBRAFISH, ECOLI_STRAIN_K12, ECOLI_STRAIN_SAKAI, CHICKEN, RHESUS, MALARIA, CHIMP, RAT, YEAST, PIG, XENOPUS
  plots:
    volcano: 1
    heatmap: 1
    LFC : -1.5 1.5 # Range of minimal log2fc
    FDR : 0.05
    heatmap_cluster_cols : 0
    heatmap_display : log2FC # log2FC or pvalue
```

Extra actions to perform based on the MSstats results, including *annotations* and *plots* (heatmaps and volcano plots)

#### Special case: Protein fractionation

To handle protein fractionation experiments, two options need to be activated

1. The keys' file must contain an additional column named "`FractionKey`" with 
the information about fractions. For example:

**Raw.file**|**IsotopeLabelType**|**Condition**|**BioReplicate**|**Run**|**FractionKey**
:-----:|:-----:|:-----:|:-----:|:-----:|:-----:
S9524\_Fx1|L|AB|AB-1|1|1
S9524\_Fx2|L|AB|AB-1|1|2
S9524\_Fx3|L|AB|AB-1|1|3
S9524\_Fx4|L|AB|AB-1|1|4
S9524\_Fx5|L|AB|AB-1|1|5
S9524\_Fx6|L|AB|AB-1|1|6
S9524\_Fx7|L|AB|AB-1|1|7
S9524\_Fx8|L|AB|AB-1|1|8
S9524\_Fx9|L|AB|AB-1|1|9
S9524\_Fx10|L|AB|AB-1|1|10
S9525\_Fx1|L|AB|AB-2|2|1
S9525\_Fx2|L|AB|AB-2|2|2
S9525\_Fx3|L|AB|AB-2|2|3
S9525\_Fx4|L|AB|AB-2|2|4
S9525\_Fx5|L|AB|AB-2|2|5
S9525\_Fx6|L|AB|AB-2|2|6
S9525\_Fx7|L|AB|AB-2|2|7
S9525\_Fx8|L|AB|AB-2|2|8
S9525\_Fx9|L|AB|AB-2|2|9
S9525\_Fx10|L|AB|AB-2|2|10
S9526\_Fx1|L|AB|AB-3|3|1
S9526\_Fx2|L|AB|AB-3|3|2
S9526\_Fx3|L|AB|AB-3|3|3
S9526\_Fx4|L|AB|AB-3|3|4
S9526\_Fx5|L|AB|AB-3|3|5
S9526\_Fx6|L|AB|AB-3|3|6
S9526\_Fx7|L|AB|AB-3|3|7
S9526\_Fx8|L|AB|AB-3|3|8
S9526\_Fx9|L|AB|AB-3|3|9
S9526\_Fx10|L|AB|AB-3|3|10

2. Enable *fractions* in the configuration file as follow:

```
fractions: 
  enabled : 1 # 1 for protein fractions, 0 otherwise
```

#### Special case: SILAC

One of the most widely used techniques that enable relative protein 
quantification is *stable isotope labeling by amino acids in cell culture* 
(SILAC). The keys file can capture the typical SILAC experiment. 
The following example shows a SILAC experiment with two conditions, 
two biological replicates, and two technical replicates:

**RawFile**|**IsotopeLabelType**|**Condition**|**BioReplicate**|**Run**
:-----:|:-----:|:-----:|:-----:|:-----:
QE20140321-01|H|iso|iso-1|1
QE20140321-02|H|iso|iso-1|2
QE20140321-04|L|iso|iso-2|3
QE20140321-05|L|iso|iso-2|4
QE20140321-01|L|iso\_M|iso\_M-1|1
QE20140321-02|L|iso\_M|iso\_M-1|2
QE20140321-04|H|iso\_M|iso\_M-2|3
QE20140321-05|H|iso\_M|iso\_M-2|4

It is also required to activate the *silac* option in the yaml file as follows:

```
silac: 
  enabled : 1 # 1 for SILAC experiments
```


## Tips

Do you need to remember the basics of markdown? [Check out this fantastic link](https://commonmark.org/help/tutorial/index.html).
