# CREATE SAMPLE DATA FILES FOR THE PACKAGE

# library(biomaRt)
# library(bit64)
# library(circlize)
# library(cluster)
# library(corrplot)
# library(data.table)
# library(factoextra)
# library(FactoMineR)
# library(getopt)
# library(ggdendro)
# library(ggplot2)
# library(gProfileR)
# library(grid)
# library(limma)
# library(MSstats)
# library(openxlsx)
# library(org.Ag.eg.db)
# library(org.At.tair.db)
# library(org.Bt.eg.db)
# library(org.Ce.eg.db)
# library(org.Cf.eg.db)
# library(org.Dm.eg.db)
# library(org.Dr.eg.db)
# library(org.EcK12.eg.db)
# library(org.EcSakai.eg.db)
# library(org.Gg.eg.db)
# library(org.Hs.eg.db)
# library(org.Mm.eg.db)
# library(org.Mmu.eg.db)
# library(org.Pf.plasmo.db)
# library(org.Pt.eg.db)
# library(org.Rn.eg.db)
# library(org.Sc.sgd.db)
# library(org.Ss.eg.db)
# library(org.Xl.eg.db)
# library(PerformanceAnalytics)
# library(pheatmap)
# library(plotly)
# library(plyr)
# library(RColorBrewer)
# library(reshape2)
# library(seqinr)
# library(shiny)
# library(stringr)
# library(VennDiagram)
# library(yaml)
# library(graphics)
# library(grDevices)
# library(stats)
# library(utils)


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# CREATE DATA

setwd('~/github/biodavidjm/artMS/')

# GENERATE RANDOM FILE
artms_data_randomDF <- data.frame(replicate(10,sample(0:1,100,rep=TRUE)))
save(artms_data_randomDF, file = 'data/artms_data_randomDF.RData', compress = 'xz')

# PH FILES

# Reduced version of an Evidence file (generated below)
artms_data_ph_evidence <- read.delim("~/experiments/artms/ph/artms_data_ph_evidence.txt", stringsAsFactors = FALSE)
save(artms_data_ph_evidence, file = 'data/artms_data_ph_evidence.RData', compress = 'xz')

# Reduced version of the Keys file (experimental design)
artms_data_ph_keys <- read.delim("~/experiments/artms/extdata/artms_data_ph_keys.txt", stringsAsFactors = FALSE)
save(artms_data_ph_keys, file = 'data/artms_data_ph_keys.RData', compress = 'xz')

# Reduced version of the results
artms_data_ph_msstats_results <- read.delim("~/experiments/artms/extdata/artms_data_ph_msstats_results.txt", stringsAsFactors = FALSE)
save(artms_data_ph_msstats_results, file = 'data/artms_data_ph_msstats_results.RData', compress = 'xz')

# CORUM dataset
artms_data_corum_mito_database <- read.delim("inst/extdata/20170801_corum_mitoT.txt", stringsAsFactors = FALSE)
save(artms_data_corum_mito_database, file = 'data/artms_data_corum_mito_database.RData', compress = 'xz')

# CONFIGURATION FILE
artms_config <- yaml.load_file("inst/extdata/artms_config.yaml")
save(artms_config, file = 'data/artms_config.RData', compress = 'xz')


load("data/artms_config.RData")


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Testing artMS

## APMS FLUOMICS
setwd("~/experiments/artms/apms/")
contrast_file <- 'a549-PB1-contrast.txt'
artms_evidenceToMISTformat(input_file = "a549-PB1-evidence.txt", 
                           keys_file = "a549-PB1-keys.txt", 
                           quant_variable = "msint", 
                           output_file = "a549-PB1-mist.txt",
                           species = "HUMAN-FLUOMICS",
                           uniprot_dir = "~/Box Sync/db/mist/")

artms_evidenceToSaintExpressFormat(input_file = "a549-PB1-evidence.txt", 
                                   keys_file = "a549-PB1-keys.txt", 
                                   ref_proteome_file = "~/Box Sync/db/flu/fluomics-uniprot-hsa_20170516.fasta", 
                                   quant_variable = "msint", 
                                   output_file = "a549-PB1-saintexpress.txt")

artms_quantification(yaml_config_file = "resultsQuant/artms_apms_config.yaml")

setwd("~/experiments/artms/apms/resultsQuant/")
artms_analysisQuantifications(log2fc_file = "a549-PB1-results.txt", 
                              modelqc_file = "a549-PB1-results_ModelQC.txt", 
                              specie = "human", output_dir = "analysisQ", 
                              enrich = TRUE, l2fc_thres = 1, 
                              ipval = "adjpvalue")

mss <- read.delim("resultsQuant/a549-PB1-results.txt", stringsAsFactors = FALSE)
artms_volcanoPlot(mss_results_sel = mss,
                  lfc_upper = 1, 
                  lfc_lower = -1, 
                  FDR = 0.05, 
                  file_name = "a549-PB1-results-volcanoPlot.pdf", 
                  PDF = T)

## FRACTIONS
setwd('~/experiments/artms/fractions/results/ab20180402/ab20180402debug/')
evidence_file <- '~/experiments/artms/fractions/petroski-cul4-evidence.txt'
keys_file <- '~/experiments/artms/fractions/petroski-cul4-keys.txt'
contrast_file <- '~/experiments/artms/fractions/petroski-cul4-contrast.txt'
yaml_config_file <- '~/experiments/artms/fractions/results/ab20180402/config-petroski-debugging.yaml'

# Quantifications
artms_quantification(yaml_config_file = yaml_config_file)

# Analysis of Quantifications
artms_analysisQuantifications(log2fc_file = "petroski-cul4-debug-results.txt",
                              modelqc_file = "petroski-cul4-debug-results_ModelQC.txt",
                              specie = "human",
                              isPtm = "noptm",
                              enrich = TRUE,
                              output_dir = "testingARTMS",
                              isFluomics = FALSE,
                              isBackground = "nobackground",
                              mnbr = 2,
                              l2fc_thres = 1.5,
                              ipval = "adjpvalue",
                              pathogen = "nopathogen")

#-------------------------------------------------------------------------------
## CREATE THE OFFICIAL PHGLOBAL COMING WITH THE PACKAGE
evidence_file <- 'evidence.txt'
keys_file <- 'keys.txt'
contrast_file <- 'contrast.txt'

edf <- read.delim(evidence_file, stringsAsFactors = FALSE, check.names = FALSE)
kdf <- read.delim(keys_file, stringsAsFactors = FALSE, check.names = FALSE)

# Select 2 biological replicates
selectedBR <- c("qx006145", "qx006148", "qx006151", "qx006152")
edfnew <- edf[which(edf$`Raw file` %in% selectedBR),]
kdfnew <- kdf[which(kdf$RawFile %in% selectedBR), ]

# And random sampling lines
n <- round(dim(edfnew)[1]/7)
edfnew2 <- edfnew[sample(nrow(edfnew), n), ]

# print out evidence & keys
write.table(edfnew2, file = "~/experiments/artms/ph/artms_data_ph_evidence.txt", quote = FALSE, sep = "\t", row.names = FALSE, col.names = T)
write.table(kdfnew, file = "~/github/biodavidjm/artMS/inst/extdata/artms_data_ph_keys.txt", quote = FALSE, sep = "\t", row.names = FALSE, col.names = T)

# PH GLOBAL: 
setwd('~/experiments/artms/ph/')
evidence_file <- "~/experiments/artms/extdata/artms_data_ph_evidence.txt"
keys_file <- "~/experiments/artms/extdata/artms_data_ph_keys.txt"
contrast_file <- 'contrast.txt'

artms_evidenceQC(evidence_file = edfnew2, 
                 keys_file = kdfnew, 
                 prot_exp = "PH")

artms_replicatePlots(input_file = evidence_file, 
                     keys_file = keys_file, 
                     replicate_file = "reduced_replicates_plots.txt", 
                     prot_exp = "PH",
                     out_file = "ph-replicates-summary.txt")

artms_quantification("~/experiments/artms/ph/phglobalreduced/phglobal_reduced_config.yaml")

setwd('~/experiments/artms/ph/phglobalreduced/')

ph_results_wide <- artms_resultsWide(
                        results_msstats = artms_data_ph_msstats_results,
                        output_file = NULL)

summary_spectral_counts <- artms_spectralCounts(evidence_file = artms_data_ph_evidence, 
                     keys_file = artms_data_ph_keys)

evidence_anno <- artms_annotationUniprot(data = artms_data_ph_evidence,
                                          columnid = "Proteins", 
                                          sps = "human")

uniprots_anno <- artms_mapUniprot2entrezGeneName(
                  uniprotkb = unique(artms_data_ph_evidence$Proteins), 
                  specie = "human")

# The data must be annotated (Protein and Gene columns)
data_annotated <- artms_annotationUniprot(
  data = artms_data_ph_msstats_results, 
  columnid = "Protein", 
  sps = "human")
# And then the enrichment
enrich_set <- artms_enrichLog2fc(
  dataset = data_annotated, 
  specie = "human", 
  background = unique(data_annotated$Gene),
  heatmaps = TRUE)

dataset = data_annotated
specie = "human"
background = unique(data_annotated$Gene)
heatmaps = TRUE

artms_analysisQuantifications(log2fc_file = "phglobal_reduced-results.txt",
                              modelqc_file = "phglobal_reduced-results_ModelQC.txt",
                              specie = "human",
                              isPtm = "noptm",
                              enrich = FALSE,
                              output_dir = "testingARTMS",
                              mnbr = 2,
                              l2fc_thres = 1.5,
                              ipval = "pvalue")


# Debugging
log2fc_file = "phglobal-results.txt"
modelqc_file = "phglobal-results_ModelQC.txt"
specie = "human"
isPtm = "noptm"
enrich = TRUE
output_dir = "testingARTMS"
mnbr = 2
l2fc_thres = 1.5
ipval = "adjpvalue"
isBackground = "nobackground"
isFluomics = FALSE
pathogen = "nopathogen"

## PHSITES
setwd('~/experiments/artms/ph/')
artms_proteinToSiteConversion(
  evidence_file = "evidence.txt", 
  ref_proteome_file = "uniprot_canonical.fasta", 
  output_file = "phsite_evidence.txt", 
  mod_type = "ph")


# Generate the site-evidence.txt file:
artms_proteinToSiteConversion(evidence_file = '')

yaml_config_file <- '~/experiments/artms/ph/phsites/phsites_config.yaml'
artms_main(yaml_config_file = yaml_config_file)

setwd('~/experiments/artms/ph/phsites/')
log2fc_file = "phsites-results.txt"
modelqc_file = "phsites-results_ModelQC.txt"
specie = "human"
enrich = "yesenrich"
output_dir = "resultsTesting"
isFluomics = TRUE
isPtm = "yesptmph"
isBackground = "nobackground"
mnbr = 2
threshold = 1
ipval = "pvalue"
pathogen = "nopathogen"

## SILAC
evidence_file <- '~/experiments/artms/silac/RI__Endosome_Abundance_NoDrugvsDrug-evidence.txt'
keys_file <- '~/experiments/artms/silac/RI__Endosome_Abundance_NoDrugvsDrug-keys.txt'
contrast_file <- '~/experiments/artms/silac/RI__Endosome_Abundance_NoDrugvsDrug-contrasts.txt'
yaml_config_file <- '~/experiments/artms/silac/results/config-silac.yaml'

evidence2silac <- artms_SILACtoLong(evidence_file = evidence_file, output = "silac-evidence.txt")

## ABUNDANCE, technical replicates
evidence_file <- '~/experiments/artms/technical_replicas/201706-FLU-HTBE-H5N1-AB-evidence.txt'
keys_file <- '~/experiments/artms/technical_replicas/FLU-HTBE-H5N1-AB-keys.txt'
contrast_file <- '~/experiments/artms/technical_replicas/FLU-HTBE-H5N1-contrasts-final.txt'
yaml_config_file <- '~/experiments/artms/technical_replicas/configTR.yaml'

# Abundance, no technical replicates
evidence_file <- '~/experiments/artms/thp1_ab_h1n1/FLU-THP1-H1N1-AB-evidence.txt'
keys_file <- '~/experiments/artms/thp1_ab_h1n1/FLU-THP1-H1N1-AB-keys.txt'
contrast_file <- '~/experiments/artms/thp1_ab_h1n1/FLU-THP1-H1N1-AB-contrasts.txt'

## ANALYSIS OF QUANTIFICATIONS
setwd('~/experiments/artms/thp1_ab_h1n1/results/testing/')

artms_analysisQuantifications(log2fc_file = "ab-testing-new-results.txt",
                              modelqc_file = "ab-testing-new-results_ModelQC.txt",
                              specie = "human",
                              isPtm = "noptm",
                              enrich = TRUE,
                              output_dir = "AnalysisQuantifications",
                              isFluomics = TRUE,
                              isBackground = "nobackground",
                              mnbr = 2,
                              l2fc_thres = 1,
                              ipval = "pvalue",
                              pathogen = "nopathogen")

q <- resultsHeatmap(results_file = "ab-testing-new-results.txt", 
                    save_file = "whatever.pdf",
                    specie = "human")
print(q)

artms_resultsWide(evidence_file = "results/testing/ab-testing-new-results.txt", 
                  output_file = "results/testing/ab-testing-new-results-wide.txt")

artms_dataPlots(input_file = "results/testing/ab-testing-new-results-mss-normalized.txt", 
                output_file = "results/testing/ab-testing-new-results-mss-normalized.pdf")

artms_plotHeatmapQuant(input_file = "ab-testing-new-results.txt", 
                  output_file = "ab-testing2-new-results-heatmap.pdf", 
                  specie = "human")

here <- artms_msstats_summary(evidence_file = "FLU-THP1-H1N1-AB-evidence.txt", 
                prot_group_file = "proteinGroups.txt", 
                keys_file = "FLU-THP1-H1N1-AB-keys.txt", 
                results_file = "results/testing/ab-testing-new-results.txt", 
                return_df = TRUE)

artms_spectralCounts(evidence_file = "FLU-THP1-H1N1-AB-evidence.txt", 
                     keys_file = "FLU-THP1-H1N1-AB-keys.txt", 
                     output_file = "FLU-THP1-H1N1-AB-spectral_counts.txt")

evidence <- read.delim("FLU-THP1-H1N1-AB-evidence.txt", stringsAsFactors = FALSE)
keys <- read.delim("FLU-THP1-H1N1-AB-keys.txt", stringsAsFactors = FALSE)
evidenceKeys <- artms_mergeEvidenceAndKeys(data = evidence, keys = keys)

evidenceKeys <- artms_mergeEvidenceKeysByFiles(evidence_file = "FLU-THP1-H1N1-AB-evidence.txt", keys_file = "FLU-THP1-H1N1-AB-keys.txt")
  
evidence_filtered <- artms_filterMaxqData(data = evidence)


evidence_file = "FLU-THP1-H1N1-AB-evidence.txt"
prot_group_file = "proteinGroups.txt"
keys_file = "FLU-THP1-H1N1-AB-keys.txt"
results_file = "results/testing/ab-testing-new-results.txt"
return_results = TRUE


log2fc_file = "ab-testing-new-results.txt"
modelqc_file = "ab-testing-new-results_ModelQC.txt"
specie = "human"
isPtm = "noptm"
enrich = TRUE
output_dir = "AnalysisQuantifications"
isFluomics = TRUE
isBackground = "nobackground"
mnbr = 2
l2fc_thres = 1
ipval = "pvalue"
pathogen = "nopathogen"


## CONFIG LOADING
# yaml_config_file <- '~/github/biodavidjm/artMS/data-raw/artms_config.yaml'
artms_main(yaml_config_file)

## Testing individual functions
here <- artms_plotHeatmap(
  input_file = '~/experiments/artms/technical_replicas/results/FLU-HTBE-H5N1-results.txt', 
  output_file = '~/experiments/artms/technical_replicas/results/FLU-HTBE-H5N1-results-plotheatmap.pdf')

## Evidence to MIST and MISTIN
artms_evidenceToMISTformat(quant_variable = "int", 
                           input_file = '~/experiments/artms/technical_replicas/201706-FLU-HTBE-H5N1-AB-evidence.txt', 
                           output_file = '~/experiments/artms/technical_replicas/201706-FLU-HTBE-H5N1-AB-evidence-mist-int.txt', 
                           keys_file = '~/experiments/artms/technical_replicas/FLU-HTBE-H5N1-AB-keys.txt', 
                           species = 'HUMAN-FLUOMICS', 
                           uniprot_dir = '~/Box Sync/db/mist/')

## Evidence QC
artms_evidenceQC(evidence_file = evidence_file, 
                 keys_file = keys_file, 
                 prot_exp = "ph")




## ANNOTATIONS
# Generate the annotation system
symbols <- c('JAK1','AATK','A2BP1','A2LD1')

select(org.Hs.eg.db, symbols, c("ENTREZID","GENENAME"), "ALIAS")

# RANDOMLY SELECT KEYS FROM UNIPROT HUMANS
uniprots <- c("Q6P996")
uniprots <- Rkeys(org.Hs.egUNIPROT)[1:100]

ano <- artms_mapUniprot2entrezGeneName(uniprotkb = uniprots, specie = "human")

library(org.Hs.eg.db)
library(org.Mm.eg.db)

# UNIPROT TO ENTREZ
uni2entrez <- select(org.Hs.eg.db, uniprots, "ENTREZID", "UNIPROT")

# Uniprot 2 symbol
mappings <- select(org.Hs.eg.db, uniprots, c("UNIPROT", "SYMBOL", "GENENAME"), keytype = "UNIPROT")
mappings <- AnnotationDbi::select(org.Hs.eg.db, uniprots, c("UNIPROT", "SYMBOL", "GENENAME", "ENTREZID"), keytype = "UNIPROT")
# Remove redundancies
mappings <- mappings[!duplicated(mappings$UNIPROT),]


# IF I NEED TO BRING ALL THE UNIPROT IDS
uni_ids <- keys(org.Hs.eg.db, c("UNIPROT"))



