\documentclass[]{article}
\usepackage{lmodern}
\usepackage{amssymb,amsmath}
\usepackage{ifxetex,ifluatex}
\usepackage{fixltx2e} % provides \textsubscript
\ifnum 0\ifxetex 1\fi\ifluatex 1\fi=0 % if pdftex
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
\else % if luatex or xelatex
  \ifxetex
    \usepackage{mathspec}
  \else
    \usepackage{fontspec}
  \fi
  \defaultfontfeatures{Ligatures=TeX,Scale=MatchLowercase}
\fi
% use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
% use microtype if available
\IfFileExists{microtype.sty}{%
\usepackage{microtype}
\UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\usepackage[margin=1in]{geometry}
\usepackage{hyperref}
\hypersetup{unicode=true,
            pdftitle={artMS:Analytical R Tools for Mass Spectrometry},
            pdfauthor={David Jimenez-Morales},
            pdfborder={0 0 0},
            breaklinks=true}
\urlstyle{same}  % don't use monospace font for urls
\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{248,248,248}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\ImportTok}[1]{#1}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\BuiltInTok}[1]{#1}
\newcommand{\ExtensionTok}[1]{#1}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.77,0.63,0.00}{#1}}
\newcommand{\RegionMarkerTok}[1]{#1}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.94,0.16,0.16}{#1}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.64,0.00,0.00}{\textbf{#1}}}
\newcommand{\NormalTok}[1]{#1}
\usepackage{graphicx,grffile}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
\IfFileExists{parskip.sty}{%
\usepackage{parskip}
}{% else
\setlength{\parindent}{0pt}
\setlength{\parskip}{6pt plus 2pt minus 1pt}
}
\setlength{\emergencystretch}{3em}  % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{0}
% Redefines (sub)paragraphs to behave more like sections
\ifx\paragraph\undefined\else
\let\oldparagraph\paragraph
\renewcommand{\paragraph}[1]{\oldparagraph{#1}\mbox{}}
\fi
\ifx\subparagraph\undefined\else
\let\oldsubparagraph\subparagraph
\renewcommand{\subparagraph}[1]{\oldsubparagraph{#1}\mbox{}}
\fi

%%% Use protect on footnotes to avoid problems with footnotes in titles
\let\rmarkdownfootnote\footnote%
\def\footnote{\protect\rmarkdownfootnote}

%%% Change title format to be more compact
\usepackage{titling}

% Create subtitle command for use in maketitle
\newcommand{\subtitle}[1]{
  \posttitle{
    \begin{center}\large#1\end{center}
    }
}

\setlength{\droptitle}{-2em}

  \title{artMS:Analytical R Tools for Mass Spectrometry}
    \pretitle{\vspace{\droptitle}\centering\huge}
  \posttitle{\par}
    \author{David Jimenez-Morales}
    \preauthor{\centering\large\emph}
  \postauthor{\par}
      \predate{\centering\large\emph}
  \postdate{\par}
    \date{2018-09-11}


\begin{document}
\maketitle

{
\setcounter{tocdepth}{2}
\tableofcontents
}
artMS provides a set of tools for the analysis of proteomics label-free
datasets.

Send your questions to
\href{mailto:artms.help@gmail.com}{\nolinkurl{artms.help@gmail.com}}

\section{Overview}\label{overview}

The functions available at \texttt{artMS} can be grouped in 4 major
categories:

\begin{itemize}
\tightlist
\item
  Quality Control
\item
  Relative quantification with MSstats
\item
  Downstream analysis of quantification (relies on the output from 2)
\item
  Miscellaneous: a series of functions to do a little bit of everything.
\end{itemize}

The quality control can be run simultaneously as part of the
\textbf{Relative Quantification} through the section \texttt{qc} of the
configuration (\texttt{yaml}) file required to run it (see details
below)

\section{Quality Control Analysis}\label{quality-control-analysis}

\textbf{Standalone version for the Quality Control analysis of the
MaxQuant evidence file}

Running a quality control of the evidence file requires two files:

\begin{itemize}
\tightlist
\item
  \texttt{evidence.txt} file obtain from MaxQuant
\item
  \texttt{keys.txt}: a file describing the experimental design of the
  evidence file
\end{itemize}

As an example, we will use a phosphorylation dataset available to
download from the \href{http://kroganlab.ucsf.edu/}{kroganlab website}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# Evidence file}
\NormalTok{url_evidence <-}\StringTok{ '~/experiments/artms/ph/evidence.txt'}
\CommentTok{# url_evidence <- 'http://kroganlab.ucsf.edu/artms/ph/evidence.txt'}
\CommentTok{# evidence.df <- read.delim(url_evidence, stringsAsFactors = F)}


\CommentTok{# Keys file}
\NormalTok{url_keys <-}\StringTok{ "~/experiments/artms/ph/keys.txt"}
\CommentTok{# url_keys <- "http://kroganlab.ucsf.edu/artms/ph/keys.txt"}
\CommentTok{# keys.df <- read.delim(url_keys, stringsAsFactors = F)}
\end{Highlighting}
\end{Shaded}

Then run \texttt{artms\_evidenceQC()} and specify the type of
phosphoproteomics experiment (argument \texttt{prot\_exp\ =\ ph}) like
this:

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{artms_evidenceQC}\NormalTok{(}\DataTypeTok{evidence_file =}\NormalTok{ url_evidence, }\DataTypeTok{keys_file =}\NormalTok{ url_keys, }\DataTypeTok{prot_exp =} \StringTok{'ph'}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

Other proteomics experiments available are \texttt{ab} (protein
abundance), \texttt{ub} (protein ubiquitination), and \texttt{apms}
(Affinity Purification Mass Spectrometry)

This function will use the \texttt{-evidence.txt} file name as a
template name to generate the output file names.

Check \texttt{artms\_evidenceQC()} to find out more about this function.

\section{Relative quantification with
MSstats}\label{relative-quantification-with-msstats}

The relative quantification is the core of this package. All the
information required to run a relative quantification analysis using
\texttt{MSstats} is provided as a configuration file (\texttt{.yaml}
format). Check the \href{../README.md}{README.md} to find out more about
the different sections of the configuration file. A template of the
configuration file is available
\href{../data-raw/artms_config.yaml}{here}.

4 different files must be provided by the user:

\begin{itemize}
\tightlist
\item
  \texttt{evidence.txt} file location
\item
  \texttt{keys.txt} file location
\item
  \texttt{contrast.txt} file location, specifying the comparisons to be
  performed
\item
  \texttt{results/results.txt} specified the file location and root name
  for the results files.
\end{itemize}

Different types of proteomics experiments can be analyzed, including
protein abundance (ab), affinity purification mass spectrometry (APMS),
and different type of posttranslational modifications, including
phosphorylation (ph), ubiquitination (ub), and acetylation (ac)

\subsection{Quantification of Changes in Protein
Abundance}\label{quantification-of-changes-in-protein-abundance}

It quantifies changes in protein abundance between two different
conditions. The template yaml file available
\href{../data-raw/artms_config.yaml}{here} is set up for protein
abundance. These are the sections to be filled up by the user:

\begin{verbatim}
files:
  evidence : /path/to/the/evidence.txt
  keys : /path/to/the/keys.txt
  contrasts : /path/to/the/contrast.txt
  output : /path/to/the/output/results_ptmGlobal/results.txt
  .
  .
  .
data:
  .
  .
  .
  filters:
    modifications : AB 
\end{verbatim}

Make sure that the filter \texttt{modifications} is labeled as
\texttt{AB}.

Finally, run the following artms function:

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{artms_quantification}\NormalTok{(}\DataTypeTok{yaml_config_file =} \StringTok{'/path/to/config/file/artms_ab_config.yaml'}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\subsection{Quantification of Changes in Global Phosphorylation /
Ubiquitination}\label{quantification-of-changes-in-global-phosphorylation-ubiquitination}

The \textbf{global phosphorylation} quantification analysis calculates
changes in phosphorylation at the \emph{protein level}. This means that
all the \textbf{modified} peptides are used to quantify changes in
protein phosphorylation. The \textbf{site-specific} quantification
(explained next) quantifies changes at the \emph{peptide level}, i.e.,
each modified peptide independently. The same applies to \textbf{global
ubiquitination} analysis

Only two sections need to be modified on the \textbf{default}
configuration (\texttt{yaml}) file:

\begin{verbatim}
files:
  evidence : /path/to/the/evidence.txt
  keys : /path/to/the/keys.txt
  contrasts : /path/to/the/contrast.txt
  output : /path/to/the/output/results_ptmGlobal/results.txt
  .
  .
  .
data:
  .
  .
  .
  filters:
    modifications : PH # Use UB for ubiquination
\end{verbatim}

The remaining options can be left unmodified.

Once the configuration \texttt{yaml} file is ready, run the following
command:

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{artms_quantification}\NormalTok{(}\DataTypeTok{yaml_config_file =} \StringTok{'/path/to/config/file/artms_phglobal_config.yaml'}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\subsection{Quantification of Changes in Site-Specific Phosphorylation /
Ubiquitination}\label{quantification-of-changes-in-site-specific-phosphorylation-ubiquitination}

The \texttt{site-specific} analysis quantifies changes at the modified
peptide level. This means that changes in every modified (ph) peptide of
a given protein will be quantified individually. The caveat is that the
proportion of missing values should increase relative to the
\textbf{global} analysis. Both \textbf{site} and \textbf{global} ph
analysis are highly correlated due to the fact that only one or two
peptides drive the overall changes in phosphorylation for every protein.
The same principle applies to \textbf{protein ubiquitination}

To run a site specific analysis follow these steps:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  A pre-processing step is required to be run on the evidence file to
  enable the site-specific ph analysis.
\end{enumerate}

For phosphorylation

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{artms_proteinToSiteConversion}\NormalTok{(}\DataTypeTok{evidence_file =} \StringTok{"/path/to/the/evidence.txt"}\NormalTok{, }\DataTypeTok{ref_proteome_file =} \StringTok{"/path/to/the/reference_proteome.fasta"}\NormalTok{, }\DataTypeTok{output_file =} \StringTok{"/path/to/the/output/ph-sites-evidence.txt"}\NormalTok{, }\DataTypeTok{mod_type =} \StringTok{"PH"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

For ubiquitination

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{artms_proteinToSiteConversion}\NormalTok{(}\DataTypeTok{evidence_file =} \StringTok{"/path/to/the/evidence.txt"}\NormalTok{, }\DataTypeTok{ref_proteome_file =} \StringTok{"/path/to/the/reference_proteome.fasta"}\NormalTok{, }\DataTypeTok{output_file =} \StringTok{"/path/to/the/output/ub-sites-evidence.txt"}\NormalTok{, }\DataTypeTok{mod_type =} \StringTok{"UB"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{1}
\tightlist
\item
  Generate a new configuration file (\texttt{phsites\_config.yaml} or
  \texttt{ubsites\_config.yaml}) as explained above, but use the ``new''
  \texttt{/path/to/the/output/ph-sites-evidence.txt} (or
  \texttt{ub-sites-evidence.txt}) file instead of the original
  \texttt{evidence.txt} file.
\end{enumerate}

Once the new \texttt{yaml} file has been created, execute:

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{artms_quantification}\NormalTok{(}\DataTypeTok{yaml_config_file =} \StringTok{'/path/to/config/file/phsites_config.yaml'}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\section{Analysis of Quantifications}\label{analysis-of-quantifications}

Comprehensive analysis of the quantification obtained running
\texttt{artms\_quantification()}. It includes:

\begin{itemize}
\tightlist
\item
  Annotations
\item
  Summary files in different format (xls, txt) and shapes (long, wide)
\item
  Numerous summary plots
\item
  Enrichment analysis using Gprofiler
\item
  PCA of protein abundance
\item
  PCA of quantification
\item
  Clustering analysis
\end{itemize}

It takes as input two files generated from the previous step
(\texttt{artms\_quantification()})

\begin{itemize}
\tightlist
\item
  \texttt{-results.txt}
\item
  \texttt{-results\_ModelQC.txt}
\end{itemize}

To run this analysis

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Set as the working directory the folder with the results obtained from
  \texttt{artms\_quantification()}.
\end{enumerate}

\begin{verbatim}
setwd('~/path/to/the/results/')
\end{verbatim}

And then run the following function:

\begin{verbatim}
artms_analysisQuantifications(log2fc_file = "ab-results.txt",
                              modelqc_file = "ab-results_ModelQC.txt",
                              specie = "human",
                              isPtm = "noptm",
                              enrich = TRUE,
                              l2fc_thres = 1,
                              ipval = "adjpvalue",
                              isBackground = "nobackground",
                              output_dir = "AnalysisQuantifications",
                              isFluomics = TRUE,
                              mnbr = 2,
                              pathogen = "nopathogen")
\end{verbatim}

A few comments:

\begin{itemize}
\tightlist
\item
  \texttt{isPTM}. For both protein abundance (\texttt{ab}) and Affinity
  Purification-Mass Spectrometry (\texttt{apms}), in addition to global
  analysis of posttranslational modifications (PH and UB), analysis use
  the option \texttt{"noptm"}. For a site specific PTM analysis use
  \texttt{"yesptmsites"}.
\item
  \texttt{specie}. The host-species supported so far are
  \texttt{"human"} and \texttt{"mouse"}
\item
  \texttt{isBackground}. If \texttt{enrich\ =\ TRUE}, the user can
  provide a background gene file of gene names. Indicate the path to the
  file in this argument.
\item
  \texttt{mnbr}: Minimal Number of Biological Replicates for imputation.
  Missing values will be imputed and this argument is set to specified
  the minimal number of biological replicates that are required in at
  least one of the conditions, but for all the comparisons. For example,
  \texttt{mnbr\ =\ 2} would mean that only proteins found in \emph{at
  least} two biological replicates will be imputed. In addition, for any
  protein, in at least one of the conditions the protein should be
  identified at least in two biological replicates or it will be
  removed.
\item
  \texttt{l2fc\_thres} is the log2fc cutoff for enrichment analysis,
  absolute value, i.e., if it is set to 1, it will consider significant
  log2fc\textgreater{} +1 and log2fc \textless{} -1.
\item
  \texttt{ipval}: specify whether \texttt{pvalue} or \texttt{adjpvalue}
  should use for the analysis. The default option is \texttt{adjpvalue}
  (multiple testing correction). But if the number of biological
  replicates for a given experiment is too low (for example n = 2), then
  \texttt{pvalue} is recommended.
\end{itemize}

Please, don't hesitate to send us your questions at
\href{mailto:artms.help@gmail.com}{\nolinkurl{artms.help@gmail.com}}


\end{document}
